aya: Implement .kconfig support

davibe opened the pull request:
Implement support for external data symbols (kconfig) following libbpf
logic.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/aya-rs/aya/1017)
<!-- Reviewable:end -->

marysaka commented on aya-obj/src/btf/btf.rs:
```diff
@@ -889,8 +889,8 @@ impl Object {
                         self.symbol_offset_by_name.insert(name, symbol.address);
+                        offset = symbol.address + section_data.len() as u64;
                         section_data.extend(data);
-                        offset = symbol.address + data.len() as u64;
```
oops nice catch!

marysaka commented on test/integration-test/bpf/kconfig.bpf.c:
```diff
@@ -3,11 +3,24 @@
+  if (CONFIG_DEFAULT_HUNG_TASK_TIMEOUT != 120) {
     return XDP_DROP;
   }
 
```
Testing more than one config was something I initially wanted to do and forgot, thanks for doing that :smile:

marysaka commented on aya/src/bpf.rs:
```diff
@@ -176,10 +176,6 @@ fn compute_kconfig_definition(features: &Features) -> HashMap<String, Vec<u8>> {
-                    if raw_value.len() > 2 || raw_value.ends_with('"') {
-                        continue;
-                    }
-
```
Oh yeah I completely inverted that check here...
Could you fix it and keep it for sanity checks as malformed data is still possible?

marysaka reviewed (commented):
Overall looks pretty solid, thanks for doing this :pray: 

I would like the fix you did for the first commit to be squashed in it if possible.
It also seems that the commit (co-)author got lost in the way but I guess that's fine.

With all comments fixed, assume my r-b/approval :+1:

alessandrod commented on aya-obj/src/btf/btf.rs:
```diff
@@ -610,6 +708,14 @@ impl Btf {
                                 };
                                 e.offset = *offset as u32;
+
+                                if var.linkage == VarLinkage::Extern {
```
Please add a comment explaining that this is needed for kconfig support

alessandrod commented on aya-obj/src/btf/btf.rs:
```diff
@@ -730,6 +836,107 @@ impl Default for Btf {
+                return Ok(None);
+            }
+
+            let mut kconfig_map_index = 0;
```
What is going on here? This is overridden at line 856?

alessandrod commented on aya-obj/src/btf/btf.rs:
```diff
@@ -473,6 +539,38 @@ impl Btf {
+        &self,
+        target_var_name: &str,
+    ) -> Result<(String, Var), BtfError> {
+        for t in &self.types.types {
```
Perf wise this isn't great, as we iterate all the types for each kconfig symbol.
We should probably do one pass where we find the datasecs, then only scan those?

alessandrod commented on test/integration-test/bpf/kconfig.bpf.c:
```diff
@@ -0,0 +1,38 @@
+// clang-format on
+
+// CONFIG_BPF=y => 1
+extern unsigned int CONFIG_BPF __kconfig;
```
We should test char, short, int and long here, to test different alignments

alessandrod commented on aya/src/bpf.rs:
```diff
@@ -109,6 +114,133 @@ pub fn features() -> &'static Features {
 }
 
+lazy_static! {
+    static ref KCONFIG_DEFINITION: HashMap<String, Vec<u8>> = compute_kconfig_definition(&FEATURES);
```
Instead of this lazy static, we could have a KConfig type. Then we could have
`EbpfLoader::with_kconfig(kconfig)`. This way we don't force people who don't
use kconfig to take the extra memory hit. It's a small hit, but people are using
aya in embedded systems so we shouldn't increase mem usage if we can.

alessandrod commented on test/integration-test/src/tests/smoke.rs:
```diff
@@ -83,3 +83,22 @@ fn extension() {
 }
+
+#[test]
+fn kconfig() {
```
we should probably have a test that checks the values passed as CONFIG_?

alessandrod commented on aya-obj/src/btf/btf.rs:
```diff
@@ -821,6 +899,139 @@ impl Default for Btf {
+                    let mut external_value_opt = externs.get(&name);
+                    let empty_data = vec![0; type_size];
+
+                    if external_value_opt.is_none() && symbol.is_weak {
```
explain this? A symbol can be weak but still have a definition?

alessandrod replied:
answering myself: weak config symbols get a 0 default value.

This is needed to handle missing CONFIG_* keys.

The code should say so, and we should make sure we do this kconfig biz only for stuff actually in the `.kconfig` section

alessandrod commented on aya-obj/src/btf/btf.rs:
```diff
@@ -821,6 +899,139 @@ impl Default for Btf {
+
+            // Build a single-pass index of all VARs contained in DATASEC types:
+            // var_name -> (datasec_name, var)
+            let mut datasec_var_index: HashMap<String, (String, Var)> = HashMap::new();
```
no need to specify the type let inference infer

davibe replied:
done

alessandrod commented on aya-obj/src/btf/btf.rs:
```diff
@@ -821,6 +899,139 @@ impl Default for Btf {
+                .filter(|(_, s)| s.name.is_some() && s.section_index.is_none() && s.is_external)
+                .map(|(_, s)| (s.name.as_ref().unwrap().clone(), s));
+
+            let mut section_data = Vec::<u8>::new();
```
nit: Vec::new()

don't specify types where they're not necessary

alessandrod commented on aya-obj/src/btf/btf.rs:
```diff
@@ -821,6 +899,139 @@ impl Default for Btf {
 }
 
 impl Object {
+    fn patch_extern_data_internal(
```
this name is confusing, maybe prepare_kconfig_section or something?

even the patch_extern_data name below is confusing. Patch makes me think you're
relocating but you're not, you're patching _the symbols_ and preparing the
section. 

Not sure what the best name would be.

alessandrod reviewed (commented):
some more comments

alessandrod commented on aya/src/bpf.rs:
```diff
@@ -94,6 +137,133 @@ pub fn features() -> &'static Features {
+        result.insert(
+            "LINUX_KERNEL_VERSION".to_string(),
+            {
+                let value = (u64::from(major) << 16) + (u64::from(minor) << 8) + u64::from(patch);
```
See KernelVersion::code

davibe replied:
done

alessandrod commented on test/integration-test/bpf/kconfig.bpf.c:
```diff
@@ -0,0 +1,48 @@
+    return XDP_DROP;
+  }
+
+  if (CONFIG_DEFAULT_HUNG_TASK_TIMEOUT != 120) {
```
is this getting tested anywhere? that the thing returns DROP if the values are wrong/corrupted?

davibe commented on aya-obj/src/btf/btf.rs:
```diff
@@ -821,6 +899,139 @@ impl Default for Btf {
+                }
+            }
+
+            if has_extern_data {
```
done

davibe commented on test/integration-test/bpf/kconfig.bpf.c:
```diff
@@ -0,0 +1,48 @@
+      return XDP_DROP;
+    }
+  }
+
```
The tests are based on known kernel configs of the kernel we selected. I think this is ok.

alessandrod commented on aya-obj/src/btf/btf.rs:
```diff
@@ -811,6 +891,206 @@ impl Default for Btf {
+                .section_infos
+                .values()
+                .map(|(index, _)| index.0)
+                .chain(self.maps.values().map(Map::section_index))
```
what is going on here? why do we need all these chains? why isn't section_infos enough?

alessandrod replied:
&Vec is always a smell, this should be &[u8]

davibe replied:
done

alessandrod reviewed (changes requested):
generally looks ok, I think the aya-obj code can be unspaghettified

davibe commented on aya/src/sys/bpf.rs:
```diff
@@ -1029,6 +1029,50 @@ pub(crate) fn is_prog_id_supported(map_type: bpf_map_type) -> bool {
+    false
+}
+
+const fn candidate_sys_bpf_symbols() -> &'static [&'static str] {
```
no, i can remove

davibe commented on aya/src/bpf.rs:
```diff
@@ -61,6 +63,45 @@ pub use aya_obj::maps::{PinningType, bpf_map_def};
+    data: HashMap<String, Vec<u8>>,
+}
+
+impl KConfig {
```
ok

davibe commented on aya-obj/src/btf/btf.rs:
```diff
@@ -811,6 +891,206 @@ impl Default for Btf {
+                .unwrap_or(0)
+                + 1;
+
+            let symbols = self
```
Agree

davibe commented on aya-obj/src/btf/btf.rs:
```diff
@@ -811,6 +891,206 @@ impl Default for Btf {
+            };
+
+            for (name, symbol) in symbols {
+                let Some((datasec_name, var)) = datasec_var_index.get(&name) else {
```
I rearranged the code differently, but thx

davibe commented on aya-obj/src/btf/btf.rs:
```diff
@@ -160,6 +161,24 @@ pub enum BtfError {
+        symbol_name: String,
+    },
+
+    /// BTF map wrapper's layout is unexpected
```
I remove

davibe commented on aya-obj/src/btf/btf.rs:
```diff
@@ -811,6 +891,206 @@ impl Default for Btf {
+                return Ok(None);
+            }
+
+            // Build a single-pass index of all VARs contained in DATASEC types:
```
yes, ok

davibe commented on aya-obj/src/btf/btf.rs:
```diff
@@ -811,6 +891,206 @@ impl Default for Btf {
+            let mut kconfig_data = Vec::new();
+            let mut offset = 0u64;
+            let endianness = self.endianness;
+            let adjust_kconfig_value = |value: &[u8], type_size: usize| -> Vec<u8> {
```
ok

davibe commented on aya-obj/src/btf/btf.rs:
```diff
@@ -811,6 +891,206 @@ impl Default for Btf {
+                value[..type_size].to_vec()
+            };
+
+            for (name, symbol) in symbols {
```
done

davibe commented on aya/Cargo.toml:
```diff
@@ -21,6 +21,7 @@ assert_matches = { workspace = true }
 aya-obj = { path = "../aya-obj", version = "^0.2.2", features = ["std"] }
 bitflags = { workspace = true }
 bytes = { workspace = true }
+flate2 = { workspace = true, features = ["rust_backend"] }
```
done

davibe commented on Cargo.toml:
```diff
@@ -77,6 +77,7 @@ dialoguer = { version = "0.12", default-features = false }
 diff = { version = "0.1.13", default-features = false }
 env_logger = { version = "0.11", default-features = false }
 epoll = { version = "4.3.3", default-features = false }
+flate2 = { version = "1.0", default-features = false }
```
done

alessandrod commented on aya/src/bpf.rs:
```diff
@@ -93,6 +157,158 @@ pub fn features() -> &'static Features {
+    }
+}
+
+fn parse_kconfig_numeric(raw_value: &str) -> Option<u64> {
```
what about signed values? -1?

davibe replied:
done

alessandrod commented on aya-obj/src/btf/btf.rs:
```diff
@@ -646,6 +713,15 @@ impl Btf {
                                 e.offset = *offset as u32;
+
+                                // For kconfig support
+                                if is_kconfig_section && var.linkage == VarLinkage::Extern {
```
now we are in the if `d.size == 0` branch, what about `> 0`? we're not fixing up
the kconfig section

davibe replied:
kconfig ones always size == 0 but yea, it's clearer to always do this

alessandrod commented on aya-obj/src/btf/btf.rs:
```diff
@@ -811,6 +887,226 @@ impl Default for Btf {
+                return Ok(None);
+            }
+
+            let mut kconfig_var_index = HashMap::new();
```
do we need to build this? why can't we do everything in a for loop?

davibe replied:
done

alessandrod commented on aya/src/sys/mod.rs:
```diff
@@ -243,3 +243,28 @@ impl From<Stats> for bpf_stats_type {
+
+        let release = CStr::from_ptr(v.release.as_ptr());
+
+        Ok(release.to_string_lossy().into_owned())
```
the fact that you're using to_string_lossy should tell you that this shouldn't
return a string.

I'm also not sure that this should be pub(crate). It has only one caller and
virtually all other callers explicitly should _not_ be using this.

davibe replied:
ok

alessandrod reviewed (commented):
some more comments

alessandrod commented on aya/src/bpf.rs:
```diff
@@ -93,6 +157,158 @@ pub fn features() -> &'static Features {
+fn read_kconfig_file(path: &PathBuf, gzip: bool) -> Option<String> {
+    let mut output = String::new();
+
+    let res = if gzip {
```
based

davibe replied:
done

alessandrod commented on aya/src/bpf.rs:
```diff
@@ -278,6 +278,26 @@ fn read_kconfig() -> Option<String> {
+    unsafe {
+        let mut v = std::mem::zeroed::<libc::utsname>();
+        if libc::uname(std::ptr::from_mut(&mut v)) != 0 {
+            return OsString::from("unknown");
```
why does this not return None?

davibe replied:
I removed this

alessandrod commented on aya/src/bpf.rs:
````diff
@@ -167,9 +405,31 @@ impl<'a> EbpfLoader<'a> {
+    /// # Ok::<(), Box<dyn std::error::Error>>(())
+    /// ```
+    #[must_use]
+    pub fn kconfig(&mut self, kconfig: KConfig) -> &mut Self {
````
I'm thinking: should we load KConfig::current() if no kconfig() is passed?

davibe replied:
but how would that work, this method gets an `Option<KConfig>` ?

alessandrod replied:
yes

- method not called: we load current()
- method called with Some(k), we load k
- method called with None, no kconfig

davibe replied:
done

alessandrod commented on aya/src/programs/mod.rs:
```diff
@@ -125,6 +125,13 @@ pub use crate::programs::{
     xdp::{Xdp, XdpError, XdpFlags},
 };
+
+pub(crate) fn test_kprobe_creation(
```
1) why is this function in between `use` statements

2) this function only has one caller and should probably not exist

3) for the is_bpf_syscall_wrapper_supported feature check we want to pass the
   current PID I think

alessandrod commented on aya-obj/src/btf/btf.rs:
```diff
@@ -811,6 +882,220 @@ impl Default for Btf {
+                        externs.get(&name).map(Vec::as_slice),
+                        self.symbol_table
+                            .get(symbol_index)
+                            .expect("extern symbol index must resolve to a symbol")
```
why are we panicking here? We have BtfError::InvalidExternalSymbol no?

alessandrod commented on aya-obj/src/btf/btf.rs:
```diff
@@ -811,6 +882,220 @@ impl Default for Btf {
+                        self.endianness,
+                    )?;
+                    let aligned_address = (offset + (type_align - 1)) & !(type_align - 1);
+                    let symbol = self
```
2nd lookup for the same value, don't

alessandrod replied:
~you're still doing two lookups~

soz my bad you fixed it

alessandrod commented on aya/src/bpf.rs:
```diff
@@ -60,6 +65,65 @@ pub use aya_obj::maps::{PinningType, bpf_map_def};
+    /// This will attempt to read `/proc/config.gz` when gzip support is enabled, otherwise
+    /// `/boot/config-<release>`, and populate the configuration with kernel version and
+    /// feature detection information.
+    pub fn current() -> Self {
```
what if neither config.gz nor config-<release> exist? this should probably fail?

alessandrod reviewed (commented):
so close

alessandrod commented on aya/src/bpf.rs:
```diff
@@ -90,6 +90,10 @@ pub struct KConfig {
 pub enum KConfigError {
+    /// No kernel config file could be found on the system.
+    #[error("kernel config not found at /proc/config.gz or /boot/config-<release>")]
+    ConfigNotFound,
```
nit: NotFound

davibe replied:
done

alessandrod commented on aya/src/bpf.rs:
```diff
@@ -92,6 +186,178 @@ pub fn features() -> &'static Features {
+        {
+            // libbpf treats unresolved weak CONFIG_ externs (including "is not set") as 0.
+            // https://github.com/torvalds/linux/blob/4334f30ebf395b204c6cbeabf371a5a998d6ba7c/tools/lib/bpf/libbpf.c#L4666-L4673
+            result.insert(format!("CONFIG_{name}"), 0u64.to_ne_bytes().to_vec());
```
ok but by doing this here irrespective of whether a symbol is strong or weak we
always end up setting commented keys as 0 right?

davibe replied:
fixed

alessandrod commented on aya-obj/src/btf/btf.rs:
```diff
@@ -811,6 +882,221 @@ impl Default for Btf {
+            return padded;
+        }
+
+        value[..type_size].to_vec()
```
I think that libbpf returns -ERANGE for this case?

And it also checks for max/min values in signed types

davibe replied:
we now raise `ExternalSymbolValueOutOfRange`

alessandrod commented on aya/src/sys/bpf.rs:
```diff
@@ -1058,6 +1058,52 @@ pub(crate) fn is_prog_id_supported(map_type: bpf_map_type) -> bool {
+    false
+}
+
+const fn candidate_sys_bpf_symbols() -> &'static [&'static str] {
```
this is missing loongarch64

we probably need to come up with a way to make this fail at compile time if we
forget an arch

davibe replied:
addressed

alessandrod reviewed (commented):
more comments

alessandrod commented on aya/src/bpf.rs:
```diff
@@ -60,6 +65,96 @@ pub use aya_obj::maps::{PinningType, bpf_map_def};
+
+    /// Creates a new `KConfig` by reading kernel configuration from the system.
+    ///
+    /// This will attempt to read `/boot/config-<release>` first, then `/proc/config.gz` when
```
Why this order? `/proc/config.gz` is supposed to be the most authoritative?

davibe replied:
I wanted to match libbpf

alessandrod commented on aya/src/sys/bpf.rs:
```diff
@@ -1058,6 +1058,67 @@ pub(crate) fn is_prog_id_supported(map_type: bpf_map_type) -> bool {
+    }
+
+    for syscall_name in SYS_BPF_SYMBOLS {
+        if create_as_probe::<KProbe>(
```
see probe::attach

this now can return the wrong result on older kernels that attach kprobes as tracepoints

alessandrod commented on aya-obj/src/btf/btf.rs:
```diff
@@ -810,7 +888,494 @@ impl Default for Btf {
+                    if value.len() < type_size {
+                        value.resize(type_size, 0);
+                    } else if value.len() > type_size {
+                        value.truncate(type_size);
```
I think we need to set *value.last_mut() = 0; here?

alessandrod reviewed (commented):
so close, yet so far

alessandrod commented:
can you squash and rebase?

alessandrod commented on aya/src/bpf.rs:
```diff
@@ -60,6 +65,96 @@ pub use aya_obj::maps::{PinningType, bpf_map_def};
+    /// This will attempt to read `/boot/config-<release>` first, then `/proc/config.gz` when
+    /// gzip support is enabled, and populate the configuration with kernel version and feature
+    /// detection information.
+    pub fn current() -> Result<Self, KConfigError> {
+        let raw_config = read_kconfig().ok_or(KConfigError::NotFound)?;
+        Ok(Self::with_raw_config(Some(&raw_config)))
+    }
```
valid, please fix this

davibe replied:
this is no longer the case, errors are errors, otherwise not found if no file is found

alessandrod commented on aya/Cargo.toml:
```diff
@@ -16,11 +16,16 @@ rust-version.workspace = true
 
+[features]
+default = ["kconfig-gz"]
+kconfig-gz = ["dep:flate2"]
```
I think this is too specific. I'd call the feature `flate2`, and gate config.gz
behind it

davibe replied:
done

alessandrod commented on aya/src/sys/bpf.rs:
```diff
@@ -1058,6 +1061,78 @@ pub(crate) fn is_prog_id_supported(map_type: bpf_map_type) -> bool {
     bpf_map_create(&mut attr).is_ok()
 }
 
+#[cfg(target_arch = "x86_64")]
```
please ask codex to double check all these values with libbpf. My codex tells me
that loongarch64 and __x32_sys_bpfre wrong

davibe replied:
done, fixed to match libbpf

alessandrod commented on aya/src/bpf.rs:
```diff
@@ -92,6 +188,186 @@ pub fn features() -> &'static Features {
+        }
+
+        let Some((key, raw_value)) = line.split_once('=') else {
+            continue;
```
why continue here? we're accepting malformed files, libbpf does not

davibe replied:
done, now they're errors

alessandrod commented on aya-obj/src/btf/btf.rs:
```diff
@@ -810,7 +888,497 @@ impl Default for Btf {
+                    let name = obj_btf.string_at(var.name_offset)?.into_owned();
+
+                    let Some(symbol_index) = external_symbols_by_name.get(name.as_str()) else {
+                        continue;
```
why are we continuing? isn't this an error?

davibe replied:
addressed, this is now treated as an error instead of being skipped

alessandrod commented on aya-obj/src/btf/btf.rs:
```diff
@@ -810,7 +888,497 @@ impl Default for Btf {
+                        key_size: size_of::<u32>() as u32,
+                        value_size: data.len() as u32,
+                        max_entries: 1,
+                        map_flags: BPF_F_RDONLY_PROG,
```
we're not setting `BPF_F_MMAPABLE` when appropriate btw (when at least a symbol
in the section is not static/hidden)

we can do this in a followup if you prefer

davibe replied:
followup

alessandrod reviewed (commented):
one more round trip then up only

davibe committed:
aya: Implement .kconfig support

davibe committed:
aya-obj: clarify kconfig type helpers

vadorovsky commented:
I'm not planning to do any stylistic review (I trust Alessandro), but the test has some issues.

On my machine I see:

```
---- tests::kconfig::kconfig stdout ----

thread 'tests::kconfig::kconfig' (2075176) panicked at test/integration-test/src/tests/kconfig.rs:124:10:
called `Result::unwrap()` on an `Err` value: BtfError(ExternalSymbolValueOutOfRange { symbol_name: "CONFIG_PANIC_TIMEOUT" })
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

 the test failure occurs because `CONFIG_PANIC_TIMEOUT` can be negative, my kernel has:

```
$ cat /boot/config-6.19.9-gentoo-dist-hardened | grep CONFIG_PANIC_TIMEOUT
CONFIG_PANIC_TIMEOUT=-1
```

But the program defines it as `unsigned short`. Kernel assigns the value of this config to an `int` variable:

https://elixir.bootlin.com/linux/v7.0/source/kernel/panic.c#L73

After fixing that I got another error:

```
---- tests::kconfig::kconfig stdout ----

thread 'tests::kconfig::kconfig' (2080828) panicked at test/integration-test/src/tests/kconfig.rs:124:10:
called `Result::unwrap()` on an `Err` value: BtfError(ExternalSymbolNotFound { symbol_name: "CONFIG_DEFAULT_HUNG_TASK_TIMEOUT" })
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

which seems undefined in my kernel:

```
$ cat /boot/config-6.19.9-gentoo-dist-hardened | grep CONFIG_DEFAULT_HUNG_TASK_TIMEOUT
$
```

So we need to declare it as `__weak`. Furthermore, the type doesn't match, it should be `unsigned long`:

https://elixir.bootlin.com/linux/v7.0/source/kernel/hung_task.c#L53

Then I got the same error as on CI:

```
---- tests::kconfig::kconfig stdout ----

thread 'tests::kconfig::kconfig' (2088652) panicked at test/integration-test/src/tests/kconfig.rs:124:10:
called `Result::unwrap()` on an `Err` value: BtfError(ExternalSymbolNotFound { symbol_name: "CONFIG_FIRST_STRING" })
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

`CONFIG_FIRST_STRING` and `CONFIG_SECOND_STRING` are your inventions not present in any kernels, which is fine for the `kconfig_unsized_strings` test, where you parse a custom config, but is not fine for the `kconfig` test that uses `KConfig::curent`. Adding `__weak` makes it work, but if you care about not having them weak for `kconfig_unsized_strings`, you'd probably need to write a separate BPF file/object.

After slapping `__weak` there, I got:

```
---- tests::kconfig::kconfig stdout ----

thread 'tests::kconfig::kconfig' (2097952) panicked at test/integration-test/src/tests/kconfig.rs:26:31:
called `Result::unwrap()` on an `Err` value: ParseIntError { kind: InvalidDigit }
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

That's the first issue I pointed out, you need to change the result type to a signed integer.

Then the undefined `CONFIG_DEFAULT_HUNG_TASK_TIMEOUT` strikes again:

```
---- tests::kconfig::kconfig stdout ----

thread 'tests::kconfig::kconfig' (2140663) panicked at test/integration-test/src/tests/kconfig.rs:30:17:
CONFIG_DEFAULT_HUNG_TASK_TIMEOUT not found in kernel config
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

Which could be fixed by returning an `Option` in `numeric_setting`.

Overall this patch makes fixes the tests:


```diff
diff --git a/test/integration-test/bpf/kconfig.bpf.c b/test/integration-test/bpf/kconfig.bpf.c
index 4bb0b50d..c2ef2595 100644
--- a/test/integration-test/bpf/kconfig.bpf.c
+++ b/test/integration-test/bpf/kconfig.bpf.c
@@ -16,19 +16,19 @@
 struct {
   __uint(type, BPF_MAP_TYPE_ARRAY);
   __type(key, __u32);
-  __type(value, __u64);
+  __type(value, int);
   __uint(max_entries, SECOND_STRING_INDEX + SECOND_STRING_LEN);
 } RESULTS SEC(".maps");

 extern _Bool CONFIG_BPF __kconfig;
-extern unsigned short CONFIG_PANIC_TIMEOUT __kconfig;
-extern unsigned int CONFIG_DEFAULT_HUNG_TASK_TIMEOUT __kconfig;
+extern int CONFIG_PANIC_TIMEOUT __kconfig;
+extern unsigned long CONFIG_DEFAULT_HUNG_TASK_TIMEOUT __kconfig __weak;
 extern _Bool CONFIG_BPF_JIT __kconfig;
 extern char CONFIG_DEFAULT_HOSTNAME[DEFAULT_HOSTNAME_LEN] __kconfig;
-extern char CONFIG_FIRST_STRING[] __kconfig;
-extern char CONFIG_SECOND_STRING[] __kconfig;
+extern char CONFIG_FIRST_STRING[] __kconfig __weak;
+extern char CONFIG_SECOND_STRING[] __kconfig __weak;

-static long set_result(__u32 key, __u64 value) {
+static long set_result(__u32 key, int value) {
   return bpf_map_update_elem(&RESULTS, &key, &value, BPF_ANY);
 }

@@ -49,7 +49,7 @@ int test_kconfig(void *ctx) {
   }
   for (int i = 0; i < DEFAULT_HOSTNAME_LEN; i++) {
     if (set_result(DEFAULT_HOSTNAME_INDEX + i,
-                   (__u64)(__u8)CONFIG_DEFAULT_HOSTNAME[i]) != 0) {
+                   (int)(__u8)CONFIG_DEFAULT_HOSTNAME[i]) != 0) {
       return 1;
     }
   }
@@ -60,13 +60,13 @@ SEC("uprobe")
 int test_kconfig_unsized_strings(void *ctx) {
   for (int i = 0; i < FIRST_STRING_LEN; i++) {
     if (set_result(FIRST_STRING_INDEX + i,
-                   (__u64)(__u8)CONFIG_FIRST_STRING[i]) != 0) {
+                   (int)(__u8)CONFIG_FIRST_STRING[i]) != 0) {
       return 1;
     }
   }
   for (int i = 0; i < SECOND_STRING_LEN; i++) {
     if (set_result(SECOND_STRING_INDEX + i,
-                   (__u64)(__u8)CONFIG_SECOND_STRING[i]) != 0) {
+                   (int)(__u8)CONFIG_SECOND_STRING[i]) != 0) {
       return 1;
     }
   }
diff --git a/test/integration-test/src/tests/kconfig.rs b/test/integration-test/src/tests/kconfig.rs
index 64ef4381..3ea9e6bb 100644
--- a/test/integration-test/src/tests/kconfig.rs
+++ b/test/integration-test/src/tests/kconfig.rs
@@ -1,4 +1,5 @@
 use std::{
+    ffi::c_int,
     fs,
     path::{Path, PathBuf},
 };
@@ -12,21 +13,24 @@ use integration_common::kconfig::{
 };
 use procfs::{ConfigSetting, KernelConfig, prelude::FromRead as _};

-fn numeric_setting(config: &std::collections::HashMap<String, ConfigSetting>, name: &str) -> u64 {
+fn numeric_setting(
+    config: &std::collections::HashMap<String, ConfigSetting>,
+    name: &str,
+) -> Option<c_int> {
     match config.get(name) {
-        Some(ConfigSetting::Yes) => 1,
-        Some(ConfigSetting::Module) => 2,
+        Some(ConfigSetting::Yes) => Some(1),
+        Some(ConfigSetting::Module) => Some(2),
         Some(ConfigSetting::Value(value)) => {
             if let Some(value) = value
                 .strip_prefix("0x")
                 .or_else(|| value.strip_prefix("0X"))
             {
-                u64::from_str_radix(value, 16).unwrap()
+                Some(c_int::from_str_radix(value, 16).unwrap())
             } else {
-                value.parse().unwrap()
+                Some(value.parse().unwrap())
             }
         }
-        None => panic!("{name} not found in kernel config"),
+        None => None,
     }
 }

@@ -130,23 +134,27 @@ fn kconfig() {

     trigger_kconfig();

-    let results: Array<_, u64> = bpf.take_map("RESULTS").unwrap().try_into().unwrap();
+    let results: Array<_, c_int> = bpf.take_map("RESULTS").unwrap().try_into().unwrap();

     assert_eq!(
         results.get(&CONFIG_BPF_INDEX, 0).unwrap(),
-        numeric_setting(&kernel_config, "CONFIG_BPF")
+        numeric_setting(&kernel_config, "CONFIG_BPF").unwrap()
     );
     assert_eq!(
         results.get(&PANIC_TIMEOUT_INDEX, 0).unwrap(),
-        numeric_setting(&kernel_config, "CONFIG_PANIC_TIMEOUT")
+        numeric_setting(&kernel_config, "CONFIG_PANIC_TIMEOUT").unwrap()
     );
-    assert_eq!(
-        results.get(&DEFAULT_HUNG_TASK_TIMEOUT_INDEX, 0).unwrap(),
+    if let Some(default_hung_task_timeout) =
         numeric_setting(&kernel_config, "CONFIG_DEFAULT_HUNG_TASK_TIMEOUT")
-    );
+    {
+        assert_eq!(
+            results.get(&DEFAULT_HUNG_TASK_TIMEOUT_INDEX, 0).unwrap(),
+            default_hung_task_timeout,
+        );
+    }
     assert_eq!(
         results.get(&BPF_JIT_INDEX, 0).unwrap(),
-        numeric_setting(&kernel_config, "CONFIG_BPF_JIT")
+        numeric_setting(&kernel_config, "CONFIG_BPF_JIT").unwrap()
     );
     let mut default_hostname = [0u8; DEFAULT_HOSTNAME_LEN];
     for (i, byte) in default_hostname.iter_mut().enumerate() {
@@ -201,7 +209,7 @@ fn kconfig_unsized_strings() {

     trigger_kconfig_unsized_strings();

-    let results: Array<_, u64> = bpf.take_map("RESULTS").unwrap().try_into().unwrap();
+    let results: Array<_, c_int> = bpf.take_map("RESULTS").unwrap().try_into().unwrap();

     let mut first = [0u8; FIRST_STRING_LEN];
     for (i, byte) in first.iter_mut().enumerate() {
```

tamird reviewed (commented):
> @tamird reviewed 4 files and all commit messages, and made 12 comments.
> *[Reviewable](https://reviewable.io/reviews/aya-rs/aya/1017)* status: 4 of 38 files reviewed, 109 unresolved discussions.

___
*[`aya/src/bpf.rs` line 10 at r1](https://reviewable.io/reviews/aya-rs/aya/1017#-Oqa8zVK8EGZvrcJdYQN:-Oqa8zVK8EGZvrcJdYQO:bz5wmkg) ([raw file](https://github.com/aya-rs/aya/blob/b82191ca55e0bd08934036557cadbcc0bb3633d1/aya/src/bpf.rs#L10)):*
> ```XML
>     sync::{Arc, LazyLock},
> };
> #[cfg(feature = "flate2")]
> ```

this cfg is checked 8 times outside of tests. can you move some things around to reduce that?
___
*[`aya/src/bpf.rs` line 85 at r1](https://reviewable.io/reviews/aya-rs/aya/1017#-Oqa97tLBX24Tmw2zdXL:-Oqa97tLBX24Tmw2zdXM:bj20rj8) ([raw file](https://github.com/aya-rs/aya/blob/b82191ca55e0bd08934036557cadbcc0bb3633d1/aya/src/bpf.rs#L85)):*
> ```XML
> /// # Ok::<(), Box<dyn std::error::Error>>(())
> /// ```
> #[derive(Debug, Clone)]
> ```

does this need to be Clone?
___
*[`aya/src/bpf.rs` line 90 at r1](https://reviewable.io/reviews/aya-rs/aya/1017#-OqaA0G08paH5aubvVJk:-Oqa9ArC1GCNwVIOgzjE:b-os60dt) ([raw file](https://github.com/aya-rs/aya/blob/b82191ca55e0bd08934036557cadbcc0bb3633d1/aya/src/bpf.rs#L90)):*
> ```XML
> }
> 
> /// The error type returned by [`KConfig::parse`].
> ```

this is not really true; Kconfig::current returns this even if it never attempts parsing
___
*[`aya/src/bpf.rs` line 94 at r1](https://reviewable.io/reviews/aya-rs/aya/1017#-Oqa9EtWBl7485TZrAE8:-Oqa9EtWBl7485TZrAE9:b81edai) ([raw file](https://github.com/aya-rs/aya/blob/b82191ca55e0bd08934036557cadbcc0bb3633d1/aya/src/bpf.rs#L94)):*
> ```XML
> pub enum KConfigError {
>     /// No kernel config file could be found on the system.
>     #[error("kernel config not found at /boot/config-<release> or /proc/config.gz")]
> ```

these paths should be properties on the variant for two reasons:
- it would be better to say precisely which path we tried rather than <release>
- it would be better to do this programmatically to avoid this error message drifting from the actual logic used
___
*[`aya/src/bpf.rs` line 106 at r1](https://reviewable.io/reviews/aya-rs/aya/1017#-Oqa9TRx0QM4Oo-dJf5p:-Oqa9TRx0QM4Oo-dJf5q:banrj1s) ([raw file](https://github.com/aya-rs/aya/blob/b82191ca55e0bd08934036557cadbcc0bb3633d1/aya/src/bpf.rs#L106)):*
> ```XML
>     },
> 
>     /// The provided kernel config contains a malformed line.
> ```

"provided" is probably not the right word?
___
*[`aya/src/bpf.rs` line 113 at r1](https://reviewable.io/reviews/aya-rs/aya/1017#-Oqa9YRn3fLVkkB-4H4K:-Oqa9YRn3fLVkkB-4H4L:b-7ropln) ([raw file](https://github.com/aya-rs/aya/blob/b82191ca55e0bd08934036557cadbcc0bb3633d1/aya/src/bpf.rs#L113)):*
> ```XML
>     },
> 
>     /// The provided kernel config is not valid UTF-8.
> ```

ditto
___
*[`aya/src/bpf.rs` line 164 at r1](https://reviewable.io/reviews/aya-rs/aya/1017#-OqaAKZK-tMjyUleF7CG:-OqaAKZK-tMjyUleF7CH:b-i5qdnu) ([raw file](https://github.com/aya-rs/aya/blob/b82191ca55e0bd08934036557cadbcc0bb3633d1/aya/src/bpf.rs#L164)):*
> ```XML
>     match mode {
>         KConfigMode::Auto => {
>             if requires_real_kconfig {
> ```

if this is only applicable to KConfigMode::Auto, should it be a property on that variant?
___
*[`aya/src/bpf.rs` line 772 at r1](https://reviewable.io/reviews/aya-rs/aya/1017#-OqaA_PG9eTryWgpI9Pq:-OqaA_PG9eTryWgpI9Pr:bl6g17z) ([raw file](https://github.com/aya-rs/aya/blob/b82191ca55e0bd08934036557cadbcc0bb3633d1/aya/src/bpf.rs#L772)):*
> ```XML
>         if let Some(kconfig) = kconfig.as_ref() {
>             obj.prepare_kconfig_section(kconfig.as_map())?;
>         }
> ```

there's a footgun here; if the user sets kconfig then `requires_real_kconfig` is a no-op. Seems like we should somehow alert the user to this problem.



_Code quote:_
```XML
            kconfig,
        } = self;
        let mut obj = Object::parse(data)?;
        obj.patch_map_data(globals.clone())?;
        let requires_real_kconfig = obj.has_config_kconfig_externs()?;
        let kconfig = resolve_kconfig(kconfig, requires_real_kconfig, KConfig::current)
            .map_err(EbpfError::KConfigError)?;
        if let Some(kconfig) = kconfig.as_ref() {
            obj.prepare_kconfig_section(kconfig.as_map())?;
        }
```
___
*[`aya/src/sys/bpf.rs` line 1120 at r1](https://reviewable.io/reviews/aya-rs/aya/1017#-OqaB0wGCqhzUDdNllEI:-OqaB0wGCqhzUDdNllEJ:b-4l2pe) ([raw file](https://github.com/aya-rs/aya/blob/b82191ca55e0bd08934036557cadbcc0bb3633d1/aya/src/sys/bpf.rs#L1120)):*
> ```XML
>             .is_ok()
>         } else {
>             create_as_trace_point::<KProbe>(
> ```

this is a bit circular, no? `create_as_probe` and `create_as_trace_point` both bottom out in `sys/bpf`, don't they? it would be great to somehow not introduce this kind of soup
___
*[`aya-obj/src/relocation.rs` line 226 at r1](https://reviewable.io/reviews/aya-rs/aya/1017#-OqaBGwa-bs38waKR4dw:-OqaBGwa-bs38waKR4dx:bvnrpbl) ([raw file](https://github.com/aya-rs/aya/blob/b82191ca55e0bd08934036557cadbcc0bb3633d1/aya-obj/src/relocation.rs#L226)):*
> ```XML
>         };
> 
>         // calls and relocation to .text symbols are handled in a separate step
> ```

update?
___
*[`aya-obj/src/relocation.rs` line 375 at r1](https://reviewable.io/reviews/aya-rs/aya/1017#-OqaBIzMC-bmhheSJ8xI:-OqaBIzMC-bmhheSJ8xJ:bvnrpbl) ([raw file](https://github.com/aya-rs/aya/blob/b82191ca55e0bd08934036557cadbcc0bb3633d1/aya-obj/src/relocation.rs#L375)):*
> ```XML
>                 })
>                 .filter(|(_rel, sym)| {
>                     // only consider text relocations, data relocations are
> ```

update?
___
*[`aya-obj/tests/fixtures/kconfig_non_tristate.bpf.o` line 0 at r1](https://reviewable.io/reviews/aya-rs/aya/1017#-OqaBaXb5NTjybkq_B6X:-OqaBaXb5NTjybkq_B6Y:beev61t) ([raw file](https://github.com/aya-rs/aya/blob/b82191ca55e0bd08934036557cadbcc0bb3633d1/aya-obj/tests/fixtures/kconfig_non_tristate.bpf.o#L0)):*
What is this? Since when are we checking in compiled objects?


<!-- Sent from Reviewable.io -->

davibe committed:
integration-test: fix kconfig coverage across kernels

davibe committed:
integration-test: move kconfig coverage end to end

davibe commented on aya/src/bpf.rs:
```diff
@@ -60,6 +62,46 @@ pub use aya_obj::maps::{PinningType, bpf_map_def};
+    ///
+    /// This will attempt to read `/proc/config.gz` or `/boot/config-<release>` and
+    /// populate the configuration with kernel version and feature detection information.
+    pub fn from_system() -> Self {
```
done

davibe committed:
aya: tighten kconfig lookup errors

davibe committed:
aya-obj: refresh relocation comments

davibe committed:
aya: move syscall wrapper probing into programs

davibe committed:
aya: reformat tests

davibe committed:
integration-test: fix merged kconfig assertions

davibe committed:
integration-test: remove dead kconfig test code

davibe committed:
aya: simplify kconfig loading

davibe committed:
aya: inline kconfig loading policy

davibe committed:
aya: tighten kconfig error docs

davibe committed:
aya: document explicit kconfig override

davibe committed:
aya: record attempted kconfig paths

davibe committed:
aya: drop KConfig Clone derive

Hold the loader's resolved kconfig as Cow<HashMap> so the Explicit arm borrows the caller's map and the Auto arms move out an owned one. KConfig no longer needs Clone.

davibe committed:
aya: fix rustfmt import order

davibe committed:
public-api: drop stale KConfig Clone snapshot

davibe reviewed (commented):
Thx!

> @davibe made 13 comments.
> *[Reviewable](https://reviewable.io/reviews/aya-rs/aya/1017#-:-OqjVjua-B_RIl8_ivfG:bb74ptu)* status: 1 of 38 files reviewed, 78 unresolved discussions (waiting on alessandrod, marysaka, and tamird).

___
*[`aya/src/bpf.rs` line 10 at r1](https://reviewable.io/reviews/aya-rs/aya/1017#-Oqa8zVK8EGZvrcJdYQN:-OqjFJ5r-aAUxrLaezpM:bgeyh42) ([raw file](https://github.com/aya-rs/aya/blob/b82191ca55e0bd08934036557cadbcc0bb3633d1/aya/src/bpf.rs#L10)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this cfg is checked 8 times outside of tests. can you move some things around to reduce that?
</blockquote></details>

I simplified the loading, mainly in 84c7f92. We're now down to 2
___
*[`aya/src/bpf.rs` line 85 at r1](https://reviewable.io/reviews/aya-rs/aya/1017#-Oqa97tLBX24Tmw2zdXL:-OqjUAjAAfBFKLSn8Gfs:b8rxrog) ([raw file](https://github.com/aya-rs/aya/blob/b82191ca55e0bd08934036557cadbcc0bb3633d1/aya/src/bpf.rs#L85)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

does this need to be Clone?
</blockquote></details>

I addressed this in in 448da15
___
*[`aya/src/bpf.rs` line 90 at r1](https://reviewable.io/reviews/aya-rs/aya/1017#-OqaA0G08paH5aubvVJk:-OqjH6ONC4D1jc962h1_:b-xxoh2h) ([raw file](https://github.com/aya-rs/aya/blob/b82191ca55e0bd08934036557cadbcc0bb3633d1/aya/src/bpf.rs#L90)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this is not really true; Kconfig::current returns this even if it never attempts parsing
</blockquote></details>

true, fixed as part of 5f912af
___
*[`aya/src/bpf.rs` line 94 at r1](https://reviewable.io/reviews/aya-rs/aya/1017#-Oqa9EtWBl7485TZrAE8:-OqjRqjr6HCqR8v_G9zi:b-y7nidp) ([raw file](https://github.com/aya-rs/aya/blob/b82191ca55e0bd08934036557cadbcc0bb3633d1/aya/src/bpf.rs#L94)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

these paths should be properties on the variant for two reasons:
- it would be better to say precisely which path we tried rather than <release>
- it would be better to do this programmatically to avoid this error message drifting from the actual logic used
</blockquote></details>

I did it in 567a37e
___
*[`aya/src/bpf.rs` line 106 at r1](https://reviewable.io/reviews/aya-rs/aya/1017#-Oqa9TRx0QM4Oo-dJf5p:-OqjFhIo16KHE9JvEyZ3:btkpyb9) ([raw file](https://github.com/aya-rs/aya/blob/b82191ca55e0bd08934036557cadbcc0bb3633d1/aya/src/bpf.rs#L106)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

"provided" is probably not the right word?
</blockquote></details>

addressed in a155097
___
*[`aya/src/bpf.rs` line 113 at r1](https://reviewable.io/reviews/aya-rs/aya/1017#-Oqa9YRn3fLVkkB-4H4K:-OqjFjdk0t8I2oqttd3F:bnj7hl0) ([raw file](https://github.com/aya-rs/aya/blob/b82191ca55e0bd08934036557cadbcc0bb3633d1/aya/src/bpf.rs#L113)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

ditto
</blockquote></details>

same as above
___
*[`aya/src/bpf.rs` line 164 at r1](https://reviewable.io/reviews/aya-rs/aya/1017#-OqaAKZK-tMjyUleF7CG:-OqjJQv59eaqFWjUKvVa:b-vf2ig3) ([raw file](https://github.com/aya-rs/aya/blob/b82191ca55e0bd08934036557cadbcc0bb3633d1/aya/src/bpf.rs#L164)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

if this is only applicable to KConfigMode::Auto, should it be a property on that variant?
</blockquote></details>

I have tightened this a bit in 6e9b1bc
___
*[`aya/src/bpf.rs` line 772 at r1](https://reviewable.io/reviews/aya-rs/aya/1017#-OqaA_PG9eTryWgpI9Pq:-OqjPlHREuz-9dQ-zE1t:b-rd561u) ([raw file](https://github.com/aya-rs/aya/blob/b82191ca55e0bd08934036557cadbcc0bb3633d1/aya/src/bpf.rs#L772)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

there's a footgun here; if the user sets kconfig then `requires_real_kconfig` is a no-op. Seems like we should somehow alert the user to this problem.
</blockquote></details>

Yes this can be improved. I have made it clearer in the doc (6664e7fa).
We could also have explicit errors carrying out the name of the symbols that are missing but then we should use them for bot Auto and Explicit. I propose to do this in a followup pr if it's ok with you.
___
*[`aya-obj/src/relocation.rs` line 226 at r1](https://reviewable.io/reviews/aya-rs/aya/1017#-OqaBGwa-bs38waKR4dw:-Oqj9mTC0uSdOtvd4ph1:b-cl8kj4) ([raw file](https://github.com/aya-rs/aya/blob/b82191ca55e0bd08934036557cadbcc0bb3633d1/aya-obj/src/relocation.rs#L226)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

update?
</blockquote></details>

I addressed this in 2eafbc0, looks ok ? 
___
*[`aya-obj/src/relocation.rs` line 375 at r1](https://reviewable.io/reviews/aya-rs/aya/1017#-OqaBIzMC-bmhheSJ8xI:-OqjEuQO9EkbrL5FTsp0:bnj7hl0) ([raw file](https://github.com/aya-rs/aya/blob/b82191ca55e0bd08934036557cadbcc0bb3633d1/aya-obj/src/relocation.rs#L375)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

update?
</blockquote></details>

same as above
___
*[`aya/src/sys/bpf.rs` line 1120 at r1](https://reviewable.io/reviews/aya-rs/aya/1017#-OqaB0wGCqhzUDdNllEI:-OqjEyl12CiK2cT7fGMR:b-wewlvp) ([raw file](https://github.com/aya-rs/aya/blob/b82191ca55e0bd08934036557cadbcc0bb3633d1/aya/src/sys/bpf.rs#L1120)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this is a bit circular, no? `create_as_probe` and `create_as_trace_point` both bottom out in `sys/bpf`, don't they? it would be great to somehow not introduce this kind of soup
</blockquote></details>

I have refactored this to be more plain in bd8e39a
___
*[`aya-obj/tests/fixtures/kconfig_non_tristate.bpf.o` line  at r1](https://reviewable.io/reviews/aya-rs/aya/1017#-OqaBaXb5NTjybkq_B6X:-OqjF69lB-LQ5LczckAY:bb5j2q8) ([raw file](https://github.com/aya-rs/aya/blob/b82191ca55e0bd08934036557cadbcc0bb3633d1/aya-obj/tests/fixtures/kconfig_non_tristate.bpf.o#L)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

What is this? Since when are we checking in compiled objects?
</blockquote></details>

nah, I have pushed AI stuff unintentionally. That's gone, I rewrote the tests


<!-- Sent from Reviewable.io -->

davibe committed:
aya: refact kconfig read, avoid reparsing

davibe committed:
integration-test: localize kconfig layout constants

davibe committed:
aya: remove KConfig::into_map

davibe committed:
aya: fix clippy in kconfig loading

davibe committed:
public-api: restore KConfig Clone snapshot

davibe committed:
integration-test: update kconfig uprobe attach scope

davibe committed:
aya: drop unused test kernel release env hook

davibe committed:
aya: drop overlapping kconfig parse tristate test

davibe committed:
aya: clarify virtual kconfig extern comment

davibe committed:
aya: inline kconfig parsing helpers

davibe committed:
aya: drop unused test kernel_release shim

davibe committed:
integration-test: fix kconfig uprobe scope import

davibe committed:
aya-obj: rename kconfig value resize helper

davibe committed:
aya-obj: inline tristate marker helper

davibe committed:
aya-obj: rename kconfig extern materializer

davibe committed:
aya-obj: rename kconfig section builder

davibe committed:
aya-obj: inline kconfig type classification

davibe committed:
aya-obj: avoid empty tristate marker panic

davibe committed:
aya-obj: satisfy clippy in tristate marker

alessandrod commented on aya-obj/src/btf/btf.rs:
```diff
@@ -811,6 +1033,390 @@ impl Default for Btf {
+    ) -> Result<(), BtfError> {
+        // We only support external data coming from kconfig.
+        if let Some((section_index, data)) = self.build_kconfig_section(externs)? {
+            self.maps.insert(
```
in this function we insert into maps, in the called function we insert into
section_infos

why not have just one function that inserts in both places?

davibe replied:
done

alessandrod commented on aya-obj/src/btf/btf.rs:
```diff
@@ -811,6 +1033,390 @@ impl Default for Btf {
+                return Ok(None);
+            }
+
+            let kconfig_map_index = self
```
add a comment

"// we're creating a synthetic section so we need a synthetic section index too"

davibe replied:
done

alessandrod commented on aya-obj/src/btf/btf.rs:
```diff
@@ -811,6 +1033,390 @@ impl Default for Btf {
+                        d.entries[entry_index].size = data.len() as u32;
+                    }
+
+                    if kconfig_data.len() < aligned_address as usize {
```
do we need this if? resize does nothing if kconfig_data.len() == aligned_address

davibe replied:
done

alessandrod commented on aya-obj/src/btf/btf.rs:
```diff
@@ -811,6 +1033,390 @@ impl Default for Btf {
+                .collect::<HashMap<_, _>>();
+
+            let mut kconfig_data = Vec::new();
+            let mut offset = 0u64;
```
this should be called write_offset, and below you turn offset into address
"aligned_address", which should instead probably be aligned_offset. Be
consistent.

davibe replied:
done

alessandrod commented on aya-obj/src/btf/btf.rs:
```diff
@@ -811,6 +1033,390 @@ impl Default for Btf {
+            let mut kconfig_data = Vec::new();
+            let mut offset = 0u64;
+
+            for type_index in 0..obj_btf.types.types.len() {
```
this outer loop can be removed if you add a function

````
fn find_data_sec<'a>(btf: &'a Btf, name: &str) -> Result<Option<(type_index, &'a DataSec)>, BtfError> {
    ...
}
```

then you clone the returned DataSec::entries

davibe replied:
done

alessandrod commented on aya-obj/src/btf/btf.rs:
```diff
@@ -811,6 +1033,390 @@ impl Default for Btf {
+                        })?;
+
+                    let resolved_type_id = obj_btf.resolve_type(var.btf_type)?;
+                    let materialize_unsized_char_array =
```
I think this can be moved closer to where it's used? It also needs a comment
that explain what's happening and an example.

davibe replied:
I removed this. It was "ergonomic" if one wanted to use an unsized `char[]` to load CONFIG_FOO_STRING but libbpf does not allow this and we want to be compatible

alessandrod commented on aya-obj/src/btf/btf.rs:
```diff
@@ -275,6 +297,150 @@ fn add_type(header: &mut btf_header, types: &mut BtfTypes, btf_type: BtfType) ->
+
+// Clone the modifier chain above a resolved leaf type so one __kconfig VAR can
+// point at a per-symbol replacement without mutating a shared BTF type.
+fn clone_type_chain_with_replacement(
```
this function is a mess. The name is bad. It's generic sounding but it isn't generic at all. The comment is also completely useless.

Fix the name. Fix the comment to include a concrete example of what it does. 

It should probably be called materialize_char_array_something and it should
materialize inside it (right now you materialize the array right before calling
this function).

alessandrod commented on aya-obj/src/btf/btf.rs:
```diff
@@ -275,6 +297,150 @@ fn add_type(header: &mut btf_header, types: &mut BtfTypes, btf_type: BtfType) ->
+
+    Ok(obj_btf.add_type(new_type))
+}
 impl Btf {
```
missing newline here

davibe replied:
done

alessandrod commented on aya-obj/src/btf/btf.rs:
```diff
@@ -811,6 +1033,390 @@ impl Default for Btf {
+            }
+        };
+
+        let is_libbpf_tristate = match resolved_type {
```
explain what this is and how it works

davibe replied:
The BTF only tells us “this extern is an enum/enum64 with this name and size”.
libbpf uses the type name libbpf_tristate as the convention for “this enum represents kernel config tristate values”. Pre-computing it here is just a way to share the same materialization path below for both enum cases

I'll add a code-comment for this

alessandrod replied:
sounds like this could be moved to `maybe_materialize_libbpf_tristate` and if
that returns `Some` you return early.

alessandrod commented on aya-obj/src/btf/btf.rs:
```diff
@@ -275,6 +297,150 @@ fn add_type(header: &mut btf_header, types: &mut BtfTypes, btf_type: BtfType) ->
 
+// Pad or truncate numeric bytes to the target width. When shrinking from 8 bytes,
+// keep the correct half for the object endianness.
+fn resize_numeric_value(value: &[u8], type_size: usize, endianness: Endianness) -> Vec<u8> {
```
why are we truncating at all? what's the use case?

davibe replied:
here we "try to adapt" the value to what's declared in the BTF, which is the same approach of libbpf. This includes validation and truncation.
For example, when aya parses a kernel config containing `CONFIG_FOO=42` this value is parsed as a signed integer, but then the btf program may use it as `extern unsigned short CONFIG_FOO __kconfig;`

alessandrod replied:
this makes no sense to me? libbpf doesn't silently truncate. We also don't
silently truncate I think since we have the _fits() function?

Then why does this function give the impression that we truncate?

I also don't understand the rustdocs: what's a native-width byte? We don't do
smaller than 8 bit bytes. Fix the comment please.

alessandrod replied:
You have nearly 100 lines of code split in 3 functions that all do the same
endianness branching and in isolation are hard to understand.

You could probably simplify by having one function that does the one thing we
really want to do.

alessandrod commented on aya-obj/src/btf/btf.rs:
```diff
@@ -811,6 +1033,390 @@ impl Default for Btf {
+        let tristate_marker = matches!(data.as_slice(), [b'n' | b'y' | b'm']).then(|| data[0]);
+
+        let data = match resolved_type {
+            BtfType::Array(Array { array, .. }) => {
```
it's not clear at all what's going on in this match arm add comments please

davibe replied:
I rearranged a bit and added comments

alessandrod commented on aya-obj/src/btf/btf.rs:
```diff
@@ -811,6 +1033,390 @@ impl Default for Btf {
+    // Validate one __kconfig extern against its declared BTF type and materialize
+    // the bytes that should be written for it into the synthetic .kconfig section.
+    // Returns the extern's required alignment together with its final byte value.
+    fn materialize_kconfig_extern(
```
half of this function is doing tristate stuff. You sure it can't be isolated better?

davibe replied:
I tried pulling things together a bit without introducing a helper fn. See if it's better

alessandrod commented on aya/src/bpf.rs:
````diff
@@ -168,9 +410,39 @@ impl<'a> EbpfLoader<'a> {
+    /// # Ok::<(), Box<dyn std::error::Error>>(())
+    /// ```
+    #[must_use]
+    pub fn kconfig(&mut self, kconfig: Option<KConfig>) -> &mut Self {
````
thinking we could expose KConfigMode here? Seems less weird than passing None
for Disabled

davibe replied:
done

alessandrod reviewed (changes requested):
more comments

davibe committed:
aya-obj: document synthetic kconfig section index

davibe committed:
aya-obj: drop redundant kconfig resize guard

davibe committed:
aya-obj: rename kconfig layout offsets consistently

davibe committed:
aya-obj: move type-chain clone helper closer to use

davibe committed:
aya-obj: install synthetic kconfig sections in one place

davibe committed:
aya-obj: look up .kconfig datasec directly

davibe committed:
aya-obj: explain numeric kconfig resizing

davibe committed:
aya-obj: document libbpf tristate enum detection

davibe committed:
aya-obj: handle libbpf tristate enums early

davibe committed:
aya-obj: add tristate marker comment

davibe committed:
aya-obj: clarify kconfig char array materialization

davibe committed:
aya-obj: materialize unsized char array types directly

davibe committed:
aya-obj: reject unsized kconfig strings

davibe committed:
aya: expose kconfig loader mode

davibe committed:
aya-obj: use std in btf tests

alessandrod commented on aya-obj/src/btf/btf.rs:
```diff
@@ -800,7 +950,363 @@ impl Default for Btf {
+                        symbol_name: symbol_name.into(),
+                    });
+                }
+                if !scalar_value_fits(
```
you have this scalar_value_fits / scalar_value_as_u64 / resize_numeric_value
repeated in 3 different places.

Surely this should be a common function, maybe materialize_scalar or something,
which optionally checks for a range so you can implement the bool thing?

davibe replied:
ok

alessandrod commented on aya/src/bpf.rs:
```diff
@@ -61,6 +67,179 @@ pub use aya_obj::maps::{PinningType, bpf_map_def};
+    },
+
+    /// Kernel config data is not valid UTF-8.
+    #[error("kernel config is not valid UTF-8")]
```
I think that this can be folded in MalformedLine we don't need an error variant
for each implementation detail

davibe replied:
done

alessandrod reviewed (commented):
some more comments

davibe committed:
aya-obj: fold scalar kconfig materialization

davibe committed:
aya: fold invalid utf8 into malformed kconfig line

davibe committed:
aya: fix kconfig rebase fallout

alessandrod commented on aya-obj/src/btf/types.rs:
```diff
@@ -1322,11 +1322,15 @@ impl BtfType {
     }
 
+    pub(crate) const fn ptr_size() -> u32 {
+        size_of::<&()>() as u32
```
this is wrong, see determine_ptr_size in libbpf

davibe replied:
Fixed by making pointer size a property of Btf instead of BtfType, and by parsing BPF object .BTF with an explicit 8-byte pointer size, matching libbpf’s btf__set_pointer_size(obj->btf, 8). Generic Btf::parse keeps the existing host-pointer-size behavior

alessandrod commented on aya-obj/src/btf/btf.rs:
```diff
@@ -821,7 +975,350 @@ impl Default for Btf {
+                }
+
+                // tristate markers are scalar values, not valid string payloads
+                if tristate_marker.is_some() {
```
I don't get this check, so we reject strings that have the same values as
tristate markers? seems wrong?

davibe replied:
https://github.com/libbpf/libbpf/blob/v1.4.0/src/libbpf.c#L1993-L2002

unquoted `CONFIG_FOO=m` is not a string and should be rejected for `char[];` quoted `CONFIG_FOO="m"` is a string and should be accepted if it fits. We keep that distinction because quoted strings get a trailing NUL byte, so they do not look like the one-byte tristate marker

alessandrod commented on aya-obj/src/btf/btf.rs:
```diff
@@ -821,7 +975,350 @@ impl Default for Btf {
+                    && matches!(int.size, 1 | 2 | 4 | 8) =>
+            {
+                if let Some(value) = tristate_marker {
+                    if int.size != 1 {
```
so we truncate values but not extend. Interesting choice. Does libbpf do this?

davibe replied:
Here we're making sure the BTF-declared value is a char before writing the tristate marker char to it.
This mirrors libbpf https://github.com/libbpf/libbpf/blob/v1.4.0/src/libbpf.c#L1973-L2002

alessandrod commented on aya-obj/src/btf/btf.rs:
```diff
@@ -821,7 +975,350 @@ impl Default for Btf {
+            }
+            BtfType::Int(int) if int.encoding() == IntEncoding::Bool && int.size == 1 => {
+                let data = if let Some(value) = tristate_marker {
+                    if value == b'm' {
```
is this what libbpf does? this "bool only if == y" seems... weird

davibe replied:
https://github.com/libbpf/libbpf/blob/v1.4.0/src/libbpf.c#L1973-L1984

Yes, only `y/n` can be stored in a bool

alessandrod commented on test/integration-test/bpf/kconfig.bpf.c:
```diff
@@ -0,0 +1,96 @@
+  __uint(max_entries, OPTIONAL_STRING_INDEX + OPTIONAL_STRING_LEN);
+} RESULTS SEC(".maps");
+
+extern _Bool CONFIG_BPF __kconfig;
```
ask codex which tests you're missing, it'll tell you a bunch

alessandrod commented on test/integration-test/bpf/kconfig_signed_i8.bpf.c:
```diff
@@ -0,0 +1,11 @@
+extern signed char CONFIG_SIGNED_VALUE __kconfig;
+
+SEC("uprobe")
+int test_kconfig_signed_i8(void *ctx) { return CONFIG_SIGNED_VALUE != 0; }
```
why is this not in kconfig.bpf.c? what's the logic for what goes where

davibe replied:
`kconfig.bpf.c` is the shared successful-program layout: the Rust side knows its result-map ABI and uses it to assert several materialized values after the program loads and runs. The smaller `kconfig_*.bpf.c` files are separate negative fixtures with their own minimal layout; they are expected to fail during load, so there is no shared result-map ABI to preserve. Keeping `kconfig_signed_i8.bpf.c` separate lets the signed range tests fail only on the signed-i8 materialization rule, without pulling in unrelated strong externs from the successful fixture.

alessandrod commented on aya/src/bpf.rs:
```diff
@@ -951,6 +1245,42 @@ mod tests {
+
+    #[test]
+    fn test_parse_kconfig_rejects_invalid_utf8_as_malformed_line() {
+        let err = super::KConfig::parse(b"CONFIG_BROKEN=\xff\n").unwrap_err();
```
you should make this multi line to make sure that we catch the right line

davibe replied:
done

alessandrod commented on aya-obj/src/btf/btf.rs:
```diff
@@ -288,6 +310,82 @@ fn add_type(header: &mut btf_header, types: &mut BtfTypes, btf_type: BtfType) ->
+fn materialize_scalar_value(
+    symbol_name: &str,
+    value: &[u8],
+    type_size: usize,
```
this should be target_type_size?

davibe replied:
renamed

alessandrod commented on aya-obj/src/btf/btf.rs:
```diff
@@ -288,6 +310,82 @@ fn add_type(header: &mut btf_header, types: &mut BtfTypes, btf_type: BtfType) ->
+            .iter()
+            .fold(0u64, |value, byte| (value << u8::BITS) | u64::from(*byte)),
+    };
+    let bits = type_size * u8::BITS as usize;
```
this should be target_type_bits? please use good names, here you have value_bits
and bits and then you do a bunch of bitwise math and it's not really clear what
is what

davibe replied:
renamed

alessandrod commented on aya-obj/src/btf/btf.rs:
```diff
@@ -288,6 +310,82 @@ fn add_type(header: &mut btf_header, types: &mut BtfTypes, btf_type: BtfType) ->
+// Materialize a numeric value into the BTF-declared scalar width. Values parsed
+// from textual kconfig are stored as 64-bit byte arrays; when the destination is
+// narrower, only redundant sign/zero-extension bytes may be discarded.
+fn materialize_scalar_value(
```
we need comprehensive unit tests for this that cover all the cases

davibe committed:
aya-obj: use BPF pointer size for object BTF

davibe committed:
kconfig: clarify scalar target handling

davibe committed:
aya-obj: isolate libbpf tristate materialization

davibe committed:
aya-obj: test scalar kconfig materialization

davibe committed:
integration-test: cover additional kconfig cases

davibe committed:
integration-test: add kconfig bpf objects to bazel
